File: Pharmaceutical_Company_Example/Structural/Java/Flyweight.java
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;

// Flyweight class
class RegulatoryProcessFlyweight {
    private List<String> commonRegulatorySteps;

    public RegulatoryProcessFlyweight() {
        commonRegulatorySteps = new ArrayList<>();
        commonRegulatorySteps.add("Submit Trial Data");
        commonRegulatorySteps.add("Review Trial Data");
        commonRegulatorySteps.add("Submit Safety Data");
        commonRegulatorySteps.add("Review Safety Data");
        commonRegulatorySteps.add("Submit Regulatory Documents");
        commonRegulatorySteps.add("Obtain Regulatory Approval");
    }

    public List<String> getSteps() {
        return commonRegulatorySteps;
    }
}

// Flyweight Factory
class FlyweightFactory {
    private Map<String, RegulatoryProcessFlyweight> flyweights;

    public FlyweightFactory() {
        flyweights = new HashMap<>();
    }

    public RegulatoryProcessFlyweight getFlyweight(String key) {
        if (!flyweights.containsKey(key)) {
            System.out.println("Creating new flyweight for: " + key);
            RegulatoryProcessFlyweight flyweight = new RegulatoryProcessFlyweight();
            flyweights.put(key, flyweight);
        } else {
            System.out.println("Reusing existing flyweight for: " + key);
        }
        return flyweights.get(key);
    }
}

// Uses flyweights for shared regulatory processes
class Drug {
    private String name;
    private RegulatoryProcessFlyweight flyweight;
    private List<String> regulatorySteps;

    public Drug(String name, RegulatoryProcessFlyweight flyweight) {
        this.name = name;
        this.flyweight = flyweight;
        this.regulatorySteps = flyweight.getSteps();
    }

    public void displayRegulatoryProcess() {
        System.out.println("Regulatory process for " + name + ":");
        for (String step : regulatorySteps) {
            System.out.println(" - " + step);
        }
    }
}

public class FlyweightExample {
    public static void main(String[] args) {
        FlyweightFactory flyweightFactory = new FlyweightFactory();

        Drug drug1 = new Drug("Test Small Molecule 01", flyweightFactory.getFlyweight("Standard Process"));
        Drug drug2 = new Drug("Test Small Molecule 02", flyweightFactory.getFlyweight("Standard Process"));
        Drug drug3 = new Drug("Test Monoclonal Antibody 01", flyweightFactory.getFlyweight("Standard Process"));

        drug1.displayRegulatoryProcess();
        drug2.displayRegulatoryProcess();
        drug3.displayRegulatoryProcess();
    }
}

File: Pharmaceutical_Company_Example/Structural/C++/Flyweight.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>

// Flyweight class
class RegulatoryProcessFlyweight {
public:
  RegulatoryProcessFlyweight() {
    commonRegulatorySteps = {
      "Submit Trial Data", 
      "Review Trial Data", 
      "Submit Safety Data", 
      "Review Safety Data", 
      "Submit Regulatory Documents", 
      "Obtain Regulatory Approval"
    };
  }
  
  const std::vector<std::string>& getSteps() const {
    return commonRegulatorySteps;
  }
  
private:
  std::vector<std::string> commonRegulatorySteps;
};

// Flyweight Factory
class FlyweightFactory {
public:
  FlyweightFactory() {}
  
  RegulatoryProcessFlyweight* getFlyweight(const std::string& key) {
    if (flyweights.find(key) == flyweights.end()) {
      std::cout << "Creating new flyweight for: " << key << std::endl;
      flyweights[key] = new RegulatoryProcessFlyweight();
    } else {
      std::cout << "Reusing existing flyweight for: " << key << std::endl;
    }
    return flyweights[key];
  }
  
  ~FlyweightFactory() {
    // Clean up dynamically allocated flyweights
    for (auto& pair : flyweights) {
      delete pair.second;
    }
  }
  
private:
  std::unordered_map<std::string, RegulatoryProcessFlyweight*> flyweights;
};

// Uses flyweights for shared regulatory processes
class Drug {
public:
  Drug(const std::string& name, RegulatoryProcessFlyweight* flyweight) 
    : name(name), flyweight(flyweight) {
    regulatorySteps = flyweight->getSteps();
  }
  
  void displayRegulatoryProcess() const {
    std::cout << "Regulatory process for " << name << ":\n";
    for (const auto& step : regulatorySteps) {
      std::cout << " - " << step << std::endl;
    }
  }
  
private:
  std::string name;
  RegulatoryProcessFlyweight* flyweight;
  const std::vector<std::string>& regulatorySteps;
};

int main() {
  FlyweightFactory flyweightFactory;
  
  Drug drug1("Test Small Molecule 01", flyweightFactory.getFlyweight("Standard Process"));
  Drug drug2("Test Small Molecule 02", flyweightFactory.getFlyweight("Standard Process"));
  Drug drug3("Test Monoclonal Antibody 01", flyweightFactory.getFlyweight("Standard Process"));
  
  drug1.displayRegulatoryProcess();
  drug2.displayRegulatoryProcess();
  drug3.displayRegulatoryProcess();
  
  return 0;
}
